### ISSUE TITLE
[2.0a0] different names and initial values of variables with and without tf.function

### ISSUE THREAD

**Issue by llan-ml** [tensorflow/tensorflow#29442]
**System information**
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0alpha0
- Python version: 3.6.5

**Code to reproduce the issue**
<pre>
# -*- coding: utf-8 -*-
# @Author  : Lin Lan (ryan.linlan@gmail.com)

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf

tf.random.set_seed(123)

class Model(tf.keras.Model):
    def __init__(self):
        super(Model, self).__init__()
        self.dense1 = tf.keras.layers.Dense(100)
        self.dense2 = tf.keras.layers.Dense(100)
        self.dense3 = tf.keras.layers.Dense(100)

    # @tf.function
    def call(self, inputs):
        x = self.dense1(inputs)
        x = self.dense2(x)
        outputs = self.dense3(x)
        return outputs

model = Model()
model(tf.random.normal((1, 10), dtype=tf.float32))
for weight in model.trainable_weights:
    print(weight.name)
</pre>

**Other info / logs**
Without `tf.function`, the output is
<pre>
model/dense/kernel:0
model/dense/bias:0
model/dense_1/kernel:0
model/dense_1/bias:0
model/dense_2/kernel:0
model/dense_2/bias:0
</pre>

With `tf.function`, the output is
<pre>
dense/kernel:0
dense/bias:0
dense_1/kernel:0
dense_1/bias:0
dense_2/kernel:0
dense_2/bias:0
</pre>

Also, the initial values also are different with and without `tf.function`.

I guess the difference is because the code generated by `tf.function` executes in a different graph. But I think most users would expect consistent behaviors between eager and graph modes.

A workaround is to first call `model(inputs)` to initialize weights in eager mode, and then replace the method with `model.call = tf.function(model.call)`.

**Comment by mdanatg** [comment]
@fchollet @alextp Would it be possible for either tf.function to respect the outer name scopes, or for Keras to communicate the desired name scope to tf.function?

**Comment by alextp** [comment]
Names in tf2 are no longer load-bearing, so getting slightly different names for the same variables is fine. There is still checkpoint compatibility if you use tf.train.Chekpoint and tf.saved_model.save.

So I think this is not ideal but harmless.

**Comment by llan-ml** [comment]
@alextp 
- Regarding to variable names (actually, omitting parent name scope), it may make tensorboard graph display badly organized.
- Except for names, the initial values of variables are also different with and without `tf.function`, which will result in that the eager and graph modes have inconsistent behaviors with the same random seed.

**Comment by alextp** [comment]
Re the initial values, they will be the same once initializers switch to the new tf random ops (the existing ones are ~impossible to make deterministic across eager and graphs).

Re the names, tensorboard will likely change to better display tf2 graphs. Stay tuned!
